scala/compiler-shared/src/org/jetbrains/jps/incremental/scala/remote/Protocol.scala

```diff
@@ -0,0 +1,87 @@
+package org.jetbrains.jps.incremental.scala.remote
+
+
+import java.io._
+import java.util.Base64
+import java.util.zip.DeflaterOutputStream
+import java.util.zip.InflaterInputStream
+
+object Protocol {
+  private val argSeparator = "#ARG#"
+  private val emptyArg = "#STUB#"
+
+  def serializeArgs(args: Seq[String]): Array[Byte] = {
+    toBase64(compress(toBytes(mergeArgs(escapeEmptyArgs(args)))))
+  }
+
+  def deserializeArgs(data: Seq[String]): Seq[String] = {
+    val Seq(mergedEncodedArgs) = data
+    unescapeEmptyArgs(separateArgs(fromBytes(decompress(fromBase64(toBytes(mergedEncodedArgs))))))
+  }
+
+  def serializeEvent(event: Event, standalone: Boolean): Array[Byte] = {
+    adjustBase64(standalone, toBase64(compress(event.toBytes)))
+  }
+
+  def deserializeEvent(data: Array[Byte]): Event = {
+    Event.fromBytes(decompress(fromBase64(data)))
+  }
+
```

Comment:
We were doing optimizations for build time. There was no one big thing. Rather multiple small things, and this is one of them. I don't have numbers, but it was something like couple % or more of execution time of JPS, reduced down to less than 1 wrt socket communication. Most of the small messages are not compressed by much, but big ones have their sizes reduced like 7-13 times. There are 2 big things sent now: the arguments for the nailgun that contain e.g. full classpath (that for us was e.g. 400 entries) sources, etc. I merged that into one message and compressed. The other event is introduced in compile to jar part - AllGenerated with generated class files. This also reduced overhead on sockets, instead of sending many small events, we send one and it is compressed as well.

Reply:
The thing is that compression / decompression could also _reduce_ the performance, especially, considering that the I/O is local. We do need benchmarks to verify the improvement.

> Most of the small messages are not compressed by much, but big ones have their sizes reduced like 7-13 times.

But the _size_ reduction is irrelevant, it's not that we're trying to save local traffic.

> This also reduced overhead on sockets, instead of sending many small events, we send one and it is compressed as well.

Again, we do need the numbers here, and separately from the compression. It's reasonable to  modify things only if that _actually_ improves something.

In any case, it's sensible to extract that modification in a separate PR.

(BTW, we're considering using protobuf for the protocol)

Reply:
I will see if I have time for benchmarks. First thing to mention is that while we were working on that we were only working on windows. 
I was looking at jprofiler output, trying to find hotspots and then find a way to reduce time spent in them. Each change was giving an improvement. I wasn't storing those results.
IO is local, but local sockets still aren't free, their impact was noticable. We reduced both number of messages and amount of data.

Protobuf would be nice in general, though I am not sure how much would it help with performance. As I understand nailgun and how it is implemented, we s till need a string, which enforces base64. Also protobuf does not compress messages AFAIK.